/* Copyright 2023 The FedLearner Authors. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

syntax = "proto3";

import "fedlearner_webconsole/proto/common.proto";

import "google/protobuf/struct.proto";

package fedlearner_webconsole.proto;



message JobDependency {
  string source = 1;
}


message JobDefinition {
    enum JobType {
        UNSPECIFIED = 0;
        RAW_DATA = 1;
        DATA_JOIN = 2;
        PSI_DATA_JOIN = 3;
        NN_MODEL_TRANINING = 4;
        TREE_MODEL_TRAINING = 5;
        NN_MODEL_EVALUATION = 6;
        TREE_MODEL_EVALUATION = 7;
        TRANSFORMER = 8;
        ANALYZER = 9;
        CUSTOMIZED = 10;
    }

    string name = 1;
    JobType job_type = 2;
    bool is_federated = 3;
    repeated Variable variables = 4;
    repeated JobDependency dependencies = 5;
    string yaml_template = 6;
    // If true, the job's latest edition used easy editing mode
    bool easy_mode = 8;
}

message Slot {
    enum ReferenceType {
        DEFAULT = 0;
        SELF = 1;
        OTHER_JOB = 2;
        WORKFLOW = 3;
        PROJECT = 4;
        SYSTEM = 5;
        // any properties in all jobs(including self and others).
        // now limits to only name properties in frontend.
        JOB_PROPERTY = 6;
    }
    string reference = 1;

    // will be delete in the future
    string default = 2 [deprecated = true];

    string help = 3;
    ReferenceType reference_type = 4;
    string label = 5;
    google.protobuf.Value default_value = 6;

    enum ValueType{
        STRING = 0;
        NUMBER = 1;
        LIST = 2;
        OBJECT = 3;
        BOOL = 4;
        INT = 5;
    }
    ValueType value_type = 7;
}

message YamlEditorInfo {
    // {"Slot_namespace": Slot}
    map<string, Slot> slots = 1;
    string meta_yaml = 2;
    repeated Variable variables = 4;
}

message WorkflowTemplateEditorInfo {
    map<string, YamlEditorInfo> yaml_editor_infos =1;
}

message WorkflowDefinition {
    reserved 2;
    string group_alias = 1;
    repeated Variable variables = 3;
    repeated JobDefinition job_definitions = 4;
}
